drivers: led_strip: Add UART driver support for WS2812

AyushKot96 opened the pull request:
### DESCRIPTION
This PR adds a new uart driver for controlling WS2812 rgb led using tx pin.

### TESTING
Testing has been done using samples/drivers/led_strip application with board stamp_c3.
For this driver to work with esp32C3, I added uart pin inversion feature in a seperate PR [72574](https://github.com/zephyrproject-rtos/zephyr/pull/72574). 
``` console
west build -b stamp_c3 samples/drivers/led_strip -p
```

simonguinot commented on samples/drivers/led_strip/boards/stamp_c3.overlay:
```diff
@@ -0,0 +1,46 @@
+	pinctrl-0 = <&uart0_default>;
+	pinctrl-names = "default";
+	data-bits = <7>;
+    tx-invert;
```
Is that needed to have a "low -> high" transition for the start bit (instead of "low -> high") ? And to use it as the start of a data bit sent over the wire (and vice versa with the stop bit) ?

Is not `tx-invert` a property specific to STM32 UART controllers ? If yes, is this driver working with other UART controllers ?

And also the indentation above is wrong :)

thedjnK replied:
this seems far too limited for widespread use, I haven't got any mcus with a tx invert and seemingly neither does the mcu you are using this on

AyushKot96 replied:
@simonguinot Yes, it is needed because in the default UART TX line, the ideal state is high. This could be wrongly assumed to be some data by the WS2812 LED or similar ICs.
Many MCUs support UART line inversion with their HAL libraries. While writing this driver, I could only find inversion support for STM32 in Zephyr. That's why I created a separate PR [#72574](https://github.com/zephyrproject-rtos/zephyr/pull/72574/commits/fe7e05138ada5a71989bd07b35e678dac7a38e49) for supporting Esp32 uart pin inversion. NXP also has their PR [#71802](https://github.com/zephyrproject-rtos/zephyr/pull/71802) for supporting uart pin inversion merged into zephyr.

AyushKot96 replied:
@thedjnK The SPI and I2s peripherals are limited in smaller microcontrollers thats why support for uart is a good addition. Most of the mcu vensors supports uart pins inversion natively and their supports is getting added into zephyr day by day. As of now support for STM32, NXP and ESP32 mcu has already been added. And as mentioned in the description, I have used esp32c3 mcu to for testing it was working fine on my end.

simonguinot replied:
Therefore this property is mandatory. So the dependency must be expressed somewhere. If the `tx-invert` property can be made generic, then maybe you can use a DT macro in the driver's Kconfig (such as `dt_node_has_prop`) to allow the driver to be enabled only if this property is available. Or something else... But the dependency must be expressed somewhere.

In addition, you have to explain it in the driver's DT binding. For now, it is written:

```
The Uart driver should be usable as long as a zephyr Uart API driver
is available for your board. Hardware specific tuning is required
using these properties:

  - uart-baudrate
```

It is not true. Without inverting the Tx levels, the driver won't work.

I'm also waiting for a detailed explanation in the binding on how the data is serialized over the Tx UART wire by using the start/stop bits with the signal polarity inverted. And a summarized description in the commit message would also be welcome :)

AyushKot96 replied:
Yeah, it make sense to mention that this property is mandatory right now. Okay will add this.
Also, as you requested I added the comments explaining serialization in the ws2812_uart.c files. 
Explaination:

3 pulses/bits of UART are used to represent one bit of RGB data. During the development of this driver, I was using an ESP32C3 MCU with its UART baud rate set to 2.5MHz. Hence, 1 bit equals 400ns.
 For WS2812 timing details, please refer to its datasheet. 
Link: https://cdn-shop.adafruit.com/datasheets/WS2812.pdf
We can represent logic 1 with 0b110 (first two bits high,remaining one bit low), and logic 0 with 0b100 (first bit high, remaining two bits low).
 We also need to invert the polarity of UART pins because in default, the idle TX line is high, but WS2812 or similar LEDs need the idle line to be low, otherwise, it will misinterpret.
 When we make the TX line inverted, the start bit will be low to high and the stop bit will be high to low.
 Hence, the nine bits we are sending in one go will always have 1 at the start and 0 at the end, we just need to calculate middle 7 bits.
 Also, keep in mind that UART always sends the MSB first. Keeping all this data in mind, I created a lookup table to ease the calculations.
 This function is just taking three bits at a time from the 24-bit color data, i.e., color[3] array (MSB first),
 and passing those three bits into the lookup table ws2812_uart_lut[] to  retrieve the 7-bit data and fills out
the serialized output buffer.

ubieda replied:
> If the tx-invert property can be made generic, then maybe you can use a DT macro in the driver's Kconfig (such as dt_node_has_prop) to allow the driver to be enabled only if this property is available

@AyushKot96 as it was suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/72574#discussion_r1598880697, I think it makes sense to follow-up with a PR to generalize this property and then enforce it on this driver.

cx-anuj-pathak replied:
@simonguinot I think having internal pin inversion capability facilitate direct connection of led with uart tx pin. but if internal is not supported then external logic inversion circuit can be used. thus we should not make it mandatory instead mention it in the documentation.

In any case, here is the PR to make tx/rx generic properties of uart driver as mentioned in  https://github.com/zephyrproject-rtos/zephyr/pull/72574#discussion_r1598880697 https://github.com/zephyrproject-rtos/zephyr/pull/72947

CC: @ubieda

simonguinot replied:
@cx-anuj-pathak external logic inversion of the UART Tx pin also needs to be described in DT somehow, probably using the tx-invert property as well...

Anyway, from the LED point of view, we need a dependency to ensure that the ws2812_uart driver can be used. And I think it would be smart to use the `tx-invert` property. Else @AyushKot96 will have to find another way.

ubieda replied:
@AyushKot96 @cx-anuj-pathak I suggest keeping this PR simple and scoping this first version of the driver to have the `tx-invert` property as a requirement to work.

AyushKot96 replied:
@ubieda @simonguinot tx-invert property is now a requirement for this driver to work.

simonguinot commented on samples/drivers/led_strip/boards/stamp_c3.overlay:
```diff
@@ -0,0 +1,46 @@
+	current-speed = <2500000>;
+	pinctrl-0 = <&uart0_default>;
+	pinctrl-names = "default";
+	data-bits = <7>;
```
Don't we need to remove the parity bit too ?

AyushKot96 replied:
Yes, we dont need to use parity bit here. Just 7 data bits along 1 start and 1 stop bits is enough to represent 3 bits of rgb data.

simonguinot reviewed (commented):
Hello @AyushKot96,

Thanks for this driver. It is very smart.

I wonder what protocol will be used next to drive WS2812 controllers :)

Please find my early feedback below. I like the idea but I have some concerns, especially with the baud rate and start stop bits usage.

thedjnK commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -52,3 +52,11 @@ config WS2812_STRIP_RPI_PICO_PIO
+	depends on DT_HAS_WORLDSEMI_WS2812_UART_ENABLED
+	select UART if $(dt_compat_on_bus,$(DT_COMPAT_WORLDSEMI_WS2812_UART),uart)
+	help
+		Enable driver for WS2812 (and compatibles) LED strip using UART.
```
1x tab then 2x spaces for help text indent

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+/* This lookup table maps 3 bits of information into (7+startBit+stopBit = 9)-bits,*/
+/* which will be used to send ws2812 rgb led */
+static const uint8_t ws2812_uart_lut[8] = {
+	0b1011011,
+	0b0011011,
+	0b1010011,
+	0b0010011,
+	0b1011010,
+	0b0011010,
+	0b1010010,
+	0b0010010,
```
this is a gccism, should use `0x` and hex format

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+};
+
+/*
+ * Serialize an 24-bit color channel value into an equivalent sequence
```
-an +a

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
```
seems like this function can go and just use k_usleep directly?

AyushKot96 replied:
Yeah thats what I was thinking too because I used the ws2812 spi driver as a reference and this function was there.
Removing it now.

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+		uart_poll_out(cfg->uart_dev, tx[i]);
+	}
+}
+static int ws2812_strip_update_rgb(const struct device *dev,
```
newline gap

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+				return -EINVAL;
+			}
+		}
+		/*Convert 24 bits of rgb data into 8 bytes of uart frame*/
```
all comments should have a space after `/*` and before `*/`

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+	return rc;
+}
+
+static int ws2812_strip_update_channels(const struct device *dev,
+					uint8_t *channels,
+					size_t num_channels)
+{
+	LOG_ERR("update_channels not implemented");
+	return -ENOTSUP;
+}
```
function can go

AyushKot96 replied:
I dont think we should keep this api NULL because the user may use this API without checking if its NULL or not.
Also, doing so will let user know that this feature is not supported at the moment with the `error` return code `-ENOTSUP`.

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+ * Returns true if and only if cfg->px_buf is big enough to convert
+ * num_pixels RGB color values into UART frames.
+ */
+static inline bool num_pixels_ok(const struct ws2812_uart_cfg *cfg,
+				 size_t num_pixels)
+{
+	size_t nbytes;
+	bool overflow;
+
+	overflow = size_mul_overflow(num_pixels, cfg->num_colors * 8, &nbytes);
+	return !overflow && (nbytes <= cfg->px_buf_size);
+}
```
this is based on older led strip API, the length is checked automatically now

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+
+static const struct led_strip_driver_api ws2812_uart_api = {
+	.update_rgb = ws2812_strip_update_rgb,
+	.update_channels = ws2812_strip_update_channels,
```
missing length function

AyushKot96 commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,21 @@
+  is available for your board. Hardware specific tuning is required
+  using these properties:
+
+    - uart-baudrate
```
Leds similar to ws2812 timings will be good to run.

AyushKot96 replied:
There are some differences in the timing requirements of this ICs with respect to ws2812, but Ics with similar timings will work.

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,232 @@
+	0x1A, /* 0b0011010 */
+	0x52, /* 0b1010010 */
+	0x12, /* 0b0010010 */
+};
```
@AyushKot96 please consider the proposition above. I'd like to have the output signal to be configurable from DT (like for the `ws2812_spi` driver).

AyushKot96 replied:
@simonguinot Hello, I apologize for the delay. I found your idea appealing and have enhanced it by adding two additional properties: one for setting bits per RGB bit, and another for the lookup table. Could you please share your thoughts on these modifications?. Thanks

AyushKot96 commented:
@simonguinot Hi, I added the changes you requested. Can you review them? Thanks.

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,252 @@
+	for (i = 0; i < 3; i++) {
+		for (j = 0; j < 8; j++) {
+			serialized_val = (serialized_val << 1) | ((color[i]>>(7-j))&0x01);
+			if (((i*8)+j) % cfg->bits_per_rgb_bit == cfg->bits_per_rgb_bit-1) {
```
Is it correct ?

Let's say we have `bits-per-rgb-bit=<4>`, `data-bits=<6>` and 4 entries in the `lookup-table` (representing all the possible values for 2 RGB bits). So we should pick-up a value in the `lookup-table` every two RGB bits. But with the above line it is every 4 RGB bits... Am I wrong ?

AyushKot96 replied:
Yeah you are right. Thanks for picking this up. I have updated this.

simonguinot reviewed (commented):
Hi @AyushKot96,

I apologize for the delay but this driver is not easy to review and it took me a while.

I think it was correct to move the array mapping RGB bits to UART bits into the DT. But I do have a lot of comments and questions. Please find them below.

AyushKot96 commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,70 @@
+  Also, keep in mind that UART always sends the MSB first.
+  Keeping all this data in mind, I created a lookup table to ease the
+  calculations.
+  This function is just taking three bits at a time from the 24-bit color
```
Copy paste error. Removed.

AyushKot96 commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,252 @@
+			}
+		}
+		uart_data_len += ws2812_uart_ser(dev, px_buf, &color[0]);
+		px_buf += uart_data_len;
```
Yeah you are right. Thanks

AyushKot96 commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,252 @@
+		px_buf += uart_data_len;
+	}
+	ws2812_uart_tx(cfg, cfg->px_buf, uart_data_len);
+	k_usleep(51);
```
okay

AyushKot96 commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,70 @@
+      table was used:
+         lookup-table = <0x5B 0x1B 0x53
+                         0x13 0x5A 0x1A
+                         0x52 0x12>;
```
This RGB UART driver is primarily designed for use with small microcontrollers that have limited speed and memory. Therefore, it is optimized to send multiple RGB bits in a single UART transaction. However, as is often the case, optimizing for performance can come at the cost of simplicity.

cx-anuj-pathak replied:
@simonguinot Performance and simplicity topic is debatable, but the deal-breaker issue that @AyushKot96 forgot to mention is the UART speed. That is limited on MCUs as opposed to I2S/SPI.
for e.g. UART speed on MCUs with TX-Invert Property. (Typical values)
ESP32 => 5Mbps, Low end STM32L0 => e.g. L051/053 => 4Mbps.

If we try to send one-frame and or zero-frame, with 1 start + 7 data +1 stop bits then clock speed requirement will ~7.2Mbps for WS2812b (9bit make up 1.25us pulse). That will be an issue for MCU that do not support such high speed.
However with current implementation it is possible to provide a lookup table only having 0 and 1 frame value, with bits_per_rgb bit set to 9.

I have tried driver sample on nucleo_u575 with following overlay file and it work fine for me, @AyushKot96 please add this overlay file in your PR to serve as additional sample
```
#include <zephyr/dt-bindings/gpio/gpio.h>
#include <zephyr/dt-bindings/led/led.h>

#include <zephyr/dt-bindings/led/led.h>

&usart3 {
	status = "okay";
	current-speed = <7200000>;
	pinctrl-0 = < &usart3_tx_pc10 &usart3_rx_pc11 >;
	pinctrl-names = "default";
	data-bits = <7>;
	tx-invert;

	led_strip: ws2812 {
		compatible = "worldsemi,ws2812-uart";
		/* WS2812 */
		chain-length = <30>; /* arbitrary; change at will */
		color-mapping = <LED_COLOR_ID_GREEN
				 LED_COLOR_ID_RED
				 LED_COLOR_ID_BLUE>;
		bits-per-rgb-bit = <9>;		
		lookup-table = <0x7C 0x60>;
	};
};

/ {
	aliases {
		led-strip = &led_strip;
	};
};
```

Let me know if you think otherwise or we can resolve this comment.
CC: @ubieda

simonguinot replied:
> @simonguinot Performance and simplicity topic is debatable, but the deal-breaker issue that @AyushKot96 forgot to mention is the UART speed. That is limited on MCUs as opposed to I2S/SPI. for e.g. UART speed on MCUs with TX-Invert Property. (Typical values) ESP32 => 5Mbps, Low end STM32L0 => e.g. L051/053 => 4Mbps.

Hi @cx-anuj-pathak, @AyushKot96 already convinced me with the memory usage optimization.

> 
> If we try to send one-frame and or zero-frame, with 1 start + 7 data +1 stop bits then clock speed requirement will ~7.2Mbps for WS2812b (9bit make up 1.25us pulse). That will be an issue for MCU that do not support such high speed. However with current implementation it is possible to provide a lookup table only having 0 and 1 frame value, with bits_per_rgb bit set to 9.

Well, this was not the idea. The idea was to send a single RGB bit per UART frame. Only 3 or 4 bits (depending on the LED controller model and specification) of the UART frame are used to encode one RGB bit. The remaining bits are set to 0 and ignored by the LED controller. So it doesn't impact the UART speed. This allows to represent zero and one RGB frame in DT as we do for SPI. For users, it is a lot simpler to build than a `lookup-table`. But this also wastes a non negligible amount of memory, which is not great...

> 
> I have tried driver sample on nucleo_u575 with following overlay file and it work fine for me, @AyushKot96 please add this overlay file in your PR to serve as additional sample. (FYI, bytes are sent in LSB first order)

Thanks for testing.

ubieda commented:
Friendly reminder, Feature Freeze for the next Release is this upcoming friday.
Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?

simonguinot commented:
> Friendly reminder, Feature Freeze for the next Release is this upcoming friday. Do you believe this PR can be targetted to get in by then, @AyushKot96 @simonguinot?

Hi @ubieda,

I am not sure. It is possible but I am still not happy with the code readability and the driver usability (i.e. explanations to users how to configure the driver with the DT properties). Also due to the issues I discovered with my previous review, I now intend to test it with a B1414 LED controller.

So we will see but I am definitely not going to rush it.

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,51 @@
+  the start bit will be low to high and the stop bit will be high to low. Hence, the nine bits
+  we are sending in one go will always have 1 at the start and 0 at the end, we just need to
+  calculate middle 7 bits. Also, keep in mind that UART always sends the MSB first. Keeping all
+  this data in mind, I created a lookup table to ease the calculations.
```
Please try to enhance the explanation above, to make it digestible and understandable for users. It is not good enough in its current state. Unfortunately I am too busy right now to provide suggestions. This has to come from you.

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,254 @@
+	 * 110 110 110(3 UART bits per RGB bit)-> 1 1 1(RGB bits).
+	 * lookup_table[1] = 0x1B->0011011(7-data-bits)-> 1+0011011+0(start/stop bits added) ->
+	 * 100 110 110(3 UART bits per RGB bit)-> 0 1 1(RGB bits)
+	 */
```
The same goes for the big comment above.

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,254 @@
+				serialized_val = 0;
+			}
+		}
+	}
```
I think that the code above must be made more readable.

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,51 @@
+
+      For example, when setting this driver for WS2812, used 7 UART data bits and 3 bits
+      to represent single RGB data bit, below table was used:
+      lookup-table = <0x5B 0x1B 0x53 0x13 0x5A 0x1A 0x52 0x12>;
```
Please explain how to users how they can build their own `lookup-table`.

AyushKot96 replied:
Hi, now you just need to provide one-frame and zero-frame, lookup table will be created automatically.

simonguinot commented on samples/drivers/led_strip/src/main.c:
```diff
@@ -19,17 +19,20 @@ LOG_MODULE_REGISTER(main);
+#define RGB(_r, _g, _b)                                                                            \
+	{                                                                                          \
+		.r = (_r), .g = (_g), .b = (_b)                                                    \
+	}
```
Please remove all the indentation changes on this file. They are irrelevant for the ws2812-uart driver. If you want to fix the `led_strip` sample then do it with a separate commit and eventually another PR. As a rule of thumb, don't mix several changes in a single commit.

Following this rule, you should consider splitting your commit into three:

- addition of the `ws2812-uart` DT binding
- addition of the `ws2812-uart` driver
- update of the `led_strip` sample

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,254 @@
+	BUILD_ASSERT(WS2812_UART_LUT_LEN(idx) ==                                                   \
+			     1 << ((UART_DATA_BITS(idx) + 2) / BITS_PER_RGB_BIT(idx)),             \
+		     "UART lookup_table property's length != 2^((data-bits+2)/bits-per-rgb-bit)"   \
+		     ".Please check your device-tree settings");                                   \
```
Here too

simonguinot commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,254 @@
+                                                                                                   \
+	BUILD_ASSERT(WS2812_UART_LUT_LEN(idx) ==                                                   \
+			     1 << ((UART_DATA_BITS(idx) + 2) / BITS_PER_RGB_BIT(idx)),             \
+		     "UART lookup_table property's length != 2^((data-bits+2)/bits-per-rgb-bit)"   \
```
Honestly a such error message won't make any sense to reader (even to you in a couple of months) :)

simonguinot commented on drivers/led_strip/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ zephyr_library_sources_ifdef(CONFIG_APA102_STRIP        apa102.c)
 zephyr_library_sources_ifdef(CONFIG_LPD880X_STRIP       lpd880x.c)
 zephyr_library_sources_ifdef(CONFIG_WS2812_STRIP_GPIO   ws2812_gpio.c)
```
In the commit message, please use upper cases for UART and RGB. These are acronyms.

simonguinot reviewed (changes requested):
Hi @AyushKot96,

Please find my review below. Mostly I am asking you to make efforts on the readability of the code and to make the explanations clearer for users.

In addition, I am currently trying to test the driver on a B1414 LED controller. But for now it is more of a struggle :) So far I am unable to configure the UART of my boards to generate a signal with a bit pulse short enough (300 ns). The shortest I can get is 500 ns and it is not really usable.

simonguinot commented:
Sorry for the delay @AyushKot96. I'll review it next week.

simonguinot commented:
@AyushKot96 I'm on it but I'm having trouble to get it working with a strip of B1414 (WS2812 compatible) LED controllers.

@soburi @thedjnK please, can you help me with the review ? Thanks :)

simonguinot commented:
I finally managed to successfully test the ws2812_uart driver with a strip of 18 B1414 LED controllers on an NXP LPC11U68 board.

I had a lot of troubles configuring the UART controller to a 3MHz speed. I suspect this will be the case with more or less all UART controllers...

I used the DT overlay below. Note that I had to use 5 UART bits to represent 1 RGB bit because the UART controller of this MCU is unable to send a frame with less than 7 data bits.

```
/*
 * Copyright (c) 2021 Seagate Technology LLC
 *
 * SPDX-License-Identifier: Apache-2.0
 */

#include <zephyr/dt-bindings/led/led.h>

&uart4 {
        status = "okay";

        current-speed = <3000000>;
        pinctrl-0 = <&uart4_default>;
        pinctrl-names = "default";
        data-bits = <8>;
        tx-invert;

        led_strip: ws2812 {
                compatible = "everlight,b1414","worldsemi,ws2812-uart";
                status = "okay";

                /* B1414 */
                chain-length = <18>; /* arbitrary; change at will */
                color-mapping = <LED_COLOR_ID_BLUE
                                 LED_COLOR_ID_GREEN
                                 LED_COLOR_ID_RED>;
                reset-delay = <250>;

                /*
                 * Everlight B1414 LED controller;
                 *
                 * Each bit of the control signal (waveform) is described with a 1.2 us pulse:
                 * 0 bit: 300 ns high and 900 ns low.
                 * 1 bit: 900 ns high and 300 ns low.
                 *
                 * At 3 MHz, one UART bit pulse lasts 333 ns.
                 *
                 * 5 UART bits per RGB bit
                 * - RGB bit 0: 10000
                 * - RGB bit 1: 11100
                 */
                rgb-frame-per-uart-frame = <2>;
                zero-frame = <0x10>;
                one-frame = <0x1c>;
                frame-len = <5>;
        };
};
  
/ {
        aliases {
                led-strip = &led_strip;
        };
};
```

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,62 @@
+
+properties:
+  rgb-frame-per-uart-frame:
+    type: int
```
I think you should add the following enumeration to ensure that the driver won't get an unsupported value:
`enum: [1, 2, 3]`

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,62 @@
+      The UART frame value representing a logic 1.
+
+  frame-len:
+    type: int
```
Maybe an enum would be useful here too.

simonguinot reviewed (commented):
Hello @AyushKot96,

First of all, I apologize for the long delay in this review.

There are good news:
- I finally managed to successfully test the ws2812_uart driver with a strip of 18 B1414 LED controllers on an NXP LPC11U68 board. This means that the driver and its macros are flexible enough to support a different controller and UART configuration than yours, which is great.
- The description messages in `worldsemi,ws2812-uart.yaml` are much more understandable. And the new DT properties are really easier for users to configure.

And here are the negative points;
- The macro are quite a nightmare to understand and probably to maintain. I wonder if providing users with an external program/script (or something else running a build time) to generate the lookup table in DT wouldn't be a better approach. I'd like to hear/read other people's thoughts on these macros. @thedjnK @soburi @kartben please can you have a look ?
- There are still a lot of inaccuracies here and there.

Please find my review below.

thedjnK commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,62 @@
+                                     ______
+  Logic 1 frame-> 0x6 or 0b110 ->  _| 1  1 |_0_
+
+  rgb-frame-per-uart-frame -> 3 (3 RGB data bits will be send per per uart transaction)
```
*sent per UART

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+
+/*
+ * Serialize a 24-bit color channel value into an equivalent sequence
+ * of 8 bytes of UART frames,.
```
```suggestion
 * of 8 bytes of UART frames.
```

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+ * Serialize a 24-bit color channel value into an equivalent sequence
+ * of 8 bytes of UART frames,.
+ */
+static inline int ws2812_uart_ser(const struct device *dev, uint8_t *buf, uint8_t color[3])
```
use serialise/serialize in function name otherwise it looks like it's short for serial

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+ */
+static inline int ws2812_uart_ser(const struct device *dev, uint8_t *buf, uint8_t color[3])
+{
+	int i, j;
```
uint8_t? The maximum value is 3 and 8

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+static inline int ws2812_uart_ser(const struct device *dev, uint8_t *buf, uint8_t color[3])
+{
+	int i, j;
+	int buff_indx = 0;
```
make unsigned

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+
+#define WS2812_UART_DEVICE(idx)                                                                    \
+                                                                                                   \
+	BUILD_ASSERT(DT_PROP(UART_NODE(idx), tx_invert),                                           \
+		     "This driver depends on tx-invert UART property and "                         \
+		     "is not set. Please check your device-tree settings");                        \
```
I don't think this should assert, give a warning perhaps if it is not set, but it would be valid operation to use a normal UART that does not support this mode with an external single input invert logic gate

cx-anuj-pathak replied:
Looked around and unable to find a way to print warning based on condition in macro expansion.
Can you please provide some reference and guidance

thedjnK replied:
It would need to be done through cmake e.g. `message(WARNING "message here")`

cx-anuj-pathak replied:
![image](https://github.com/user-attachments/assets/1c101223-963b-4575-a5c7-6fdf55d2b17f)

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	const uint8_t *uart_lookup_table;
+	uint16_t reset_delay;
```
I don't understand why reset_delay is a uint16 when it has a maximum value of 8

cx-anuj-pathak replied:
Code is certainly confusing here.
- we have a dts prop `reset-delay` with default value 8
- then we have a MAX(dts_prop_or(reset-delay, 50), 8)

- dts_prop_or usage is redundant, because it already has 8 as default value. however minimum value is restricted to 8.
i have removed both, now reset-delay property is used as it is.

Reason for uint16_t, it is in micro-seconds, and reset delay value could easily cross 255us value for some led controllers

cx-anuj-pathak replied:
can we resolve it?

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+		.px_buf = ws2812_uart_##idx##_px_buf,                                              \
+		.num_colors = WS2812_NUM_COLORS(idx),                                              \
+		.color_mapping = ws2812_uart_##idx##_color_mapping,                                \
+		.reset_delay = MAX(WS2812_RESET_DELAY(idx), 8),                                    \
```
and this doesn't even make sense, the default above is 50, but here it is being capped at 8? How does that work?

thedjnK commented on samples/drivers/led_strip/boards/nucleo_u575zi_q.overlay:
```diff
@@ -0,0 +1,37 @@
+#include <zephyr/dt-bindings/gpio/gpio.h>
+#include <zephyr/dt-bindings/led/led.h>
+
+#include <zephyr/dt-bindings/led/led.h>
```
duplicate

matt-wood-ct commented:
This is very cool, great work man. These addressable LEDs are tricky so the more alternative drive methods the better 😄 I think the next one worth doing after UART is [Espressif's RMT](https://github.com/espressif/esp-idf/tree/master/examples/peripherals/rmt/led_strip), that is a classic and the ESP series SOCs are very popular these days.

cx-anuj-pathak commented:
@simonguinot @thedjnK 
FYI, Since @AyushKot96 is busy in some other critical task, I'll resolve all pending and future review points.
I hope you are fine with it,

cx-anuj-pathak commented on samples/drivers/led_strip/boards/nucleo_u575zi_q.overlay:
```diff
@@ -0,0 +1,37 @@
+/*
```
I agree that commit restructuring is required, but can we do it later to ease up review cycle?
for e.g. i'm planning to push all review point fixes in a separate commit so that reviewer can easily find them, and then once everything is passed i'll restructure commits.
I can do it with each review iteration too, if that is a preferred approach

simonguinot replied:
> I agree that commit restructuring is required, but can we do it later to ease up review cycle? for e.g. i'm planning to push all review point fixes in a separate commit so that reviewer can easily find them, and then once everything is passed i'll restructure commits. I can do it with each review iteration too, if that is a preferred approach

Please don't do that. Fix and amend your commits at each iteration. Reviewers will be able to track changes.

cx-anuj-pathak replied:
commits rearranged, please review

cx-anuj-pathak commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,333 @@
+		     "data-bits+2 should be a multiple of frame-len");                             \
+                                                                                                   \
+	static uint8_t ws2812_uart_##idx##_px_buf[WS2812_UART_BUFSZ(                               \
+		idx, DT_INST_PROP(idx, rgb_frame_per_uart_frame))];                                \
```
Yes

cx-anuj-pathak commented:
> And here are the negative points;
> 
> * The macro are quite a nightmare to understand and probably to maintain. I wonder if providing users with an external program/script (or something else running a build time) to generate the lookup table in DT wouldn't be a better approach. I'd like to hear/read other people's thoughts on these macros. @thedjnK @soburi @kartben please can you have a look ?

Do we have any resolution on this?
In my view macro is integrated solution with some added effort on maintaining,
However a script is multistep solution (generate and then use it in board files). I don't know if we can integrate it in the build process or not.

Having said that i can change it back to a lut property in dts, with a python script to generate lut. If community inclined to have a python script rather than having a macro in the driver, Please let me know where to place that and add documentation of that to guide the user of this driver effectively.

Note: I'm unable to resolve review points. please resolve them or let me know if i should open a new PR just to resolve review points.

cx-anuj-pathak committed:
dts: bindings: led_strip: Added WS2812-UART dts

Added worldsemi,ws2812-uart DT binding.

Signed-off-by: Ayush Kothari <ayush@croxel.com>
Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
drivers: led_strip: Added ws2812-uart driver

- ws2812 or compatible led's can be driven using uart peripheral
if internal or external tx-pin-inversion is used. We could send multiple
pixel bits in one uart frame, making it ram efficient.

Signed-off-by: Ayush Kothari <ayush@croxel.com>
Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak committed:
samples: drivers: led_strip: extented to support ws2812-uart driver

- added overlay file to test it on stamp_c3 board with on board led
- added overlay file to test it on nucleo_u575 board with external led

Signed-off-by: Ayush Kothari <ayush@croxel.com>
Signed-off-by: Anuj Pathak <anuj@croxel.com>

cx-anuj-pathak commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,62 @@
+    type: int
+    required: true
+    description: |
+      Total number of RGB bits send per UART transaction.
```
have i missed it, or this can be resolved

cx-anuj-pathak commented on dts/bindings/led_strip/worldsemi,ws2812-uart.yaml:
```diff
@@ -0,0 +1,62 @@
+
+  Calculation will be done using these node properties to build a lookup table using C macros
+  automatically. And that lookup-table will be used to feed the actual UART buffer which
+  will be send to WS2812 ic.
```
can we resolve it too?

cx-anuj-pathak commented:
@simonguinot @thedjnK gentle reminder

thedjnK commented on drivers/led_strip/CMakeLists.txt:
```diff
@@ -6,7 +6,15 @@ zephyr_library_sources_ifdef(CONFIG_APA102_STRIP        apa102.c)
 zephyr_library_sources_ifdef(CONFIG_TLC5971_STRIP       tlc5971.c)
 zephyr_library_sources_ifdef(CONFIG_TLC59731_STRIP      tlc59731.c)
+
+if (CONFIG_WS2812_STRIP_UART)
```
```suggestion
if(CONFIG_WS2812_STRIP_UART)
```

thedjnK commented on drivers/led_strip/CMakeLists.txt:
```diff
@@ -6,7 +6,15 @@ zephyr_library_sources_ifdef(CONFIG_APA102_STRIP        apa102.c)
 zephyr_library_sources_ifdef(CONFIG_TLC59731_STRIP      tlc59731.c)
+
+if (CONFIG_WS2812_STRIP_UART)
+message(WARNING "WS2812-UART driver depends on tx-pin-inversion. Please make sure \
+you either have it enabled in uart property (if supported by the driver) or \
+have external hardware setup in place to invert tx pin's logic level"
+)
```
indent with 2 spaces

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,317 @@
+ */
+#define DT_DRV_COMPAT worldsemi_ws2812_uart
+
+#include <zephyr/drivers/led_strip.h>
+
+#include <string.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_uart);
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/math_extras.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/dt-bindings/led/led.h>
```
group these includes, zephyr and non zephyr can be separated but don't need 3 sets of zephyr includes

thedjnK commented on drivers/led_strip/ws2812_uart.c:
```diff
@@ -0,0 +1,317 @@
+	const struct ws2812_uart_cfg *cfg = dev->config;
+
+	/* Loop through each of the 3 color components (Red, Green, Blue) */
+	for (uint8_t i = 0; i < 3; i++) {
```
3 is not correct, someone might define 1 colour, 3 colours or even 8, this needs to get it from dts

simonguinot reviewed (changes requested):
Hello @AyushKot96,

First thanks for taking over this PR.

Now I'm not happy with the binding description and the overall quality of the driver code. I wish Zephyr had a staging area (as in Linux). It would have been a perfect place for this driver. Indeed I think this driver could be useful and then it would be nice to merge it into Zephyr somehow. But in its current state, it is not possible.

IMHO the biggest problem is macros. They are pretty much unreadable and impossible to maintain. Can you fix that ? I am not sure it is even possible... Eventually a better solution would be to use an external program to generate the lookup table at build time. Maybe you can try to investigate this solution...

Basically you have to bring the driver into a mergeable state. I'll help you as much as I can. But I can't propose a full rewrite of the `worldsemi,ws2812-uart.yaml` binding at each review. I don't have that kind of time. Reviewers can give you directions and advice, but you have to be more proactive on the reviews.

Thanks in advance !

matt-wood-ct commented:
@cx-anuj-pathak We just tried the driver out of tree on an IMXRT1042 series SOC and ran into some issues, the driver serialised the data OK but we found that any UART config with <8 data bits left time gaps between frames (NXP peripheral hardware issue I suspect) which ruined the timing and also we found the timing was inconsistent due to stuttering between frames due to the use of the polling API.

In the end we had to operate with 8 data bits so 10 bits per frame and run the UART at 8MHz using the async DMA based API yielding consistent timing of about 360uS and 875uS which seems to work fine for our setup.  

The driver does not seem to support this setup as `frame-len` is limited to 3-9, so we ended up writing some test code in main to validate the hardware / firmware setup.  

UART config:

- tx-invert
- databits = 8
- baud rate 8mhz

We just manually generated at 24 byte buffer filled with 0xFC then we filled the first 8 bytes with 0x0C and wrote that out the uart using the asynchronous DMA based API to ensure there is not stuttering in the transmission.  
This works perfectly for us.  We will package this into a bare bones out of tree driver for the mean time, if this driver ends up supporting this arrangement and gets merged we will switch over to it.
